lib/Parse/ParseExpr.cpp

```diff
@@ -2753,17 +2753,49 @@ void Parser::ParseBlockId(SourceLocation CaretLoc) {
 }
 
 bool Parser::StartsBoundsExpression(Token &tok) {
-  if (Tok.getKind() == tok::identifier) {
+  if (tok.getKind() == tok::identifier) {
     IdentifierInfo *Ident = Tok.getIdentifierInfo();
     return (Ident == Ident_byte_count || Ident == Ident_count ||
             Ident == Ident_bounds);
   }
   return false;
 }
 
+bool Parser::StartsInteropTypeAnnotation(Token &tok) {
+  switch (tok.getKind()) {
+    case tok::kw__Ptr:
+    case tok::kw__Array_ptr:
+      return true;
+    default:
+      return false;
+  }
+}
+
+ExprResult Parser::ParseInteropTypeAnnotation() {
+  if (StartsInteropTypeAnnotation(Tok)) {
+    TypeResult Ty = ParseTypeName();
```

Comment:
Should we perform any additional inspection/validation?  See my overarching question on the PR.